gluster/subvol/cmd/gluster-subvol-provisioner/gluster-subvol-provisioner.go

```diff
@@ -0,0 +1,608 @@
+		}
+
+		for k, v := range updates {
+			result.Annotations[k] = v
+		}
+		_, updateErr := p.client.CoreV1().PersistentVolumeClaims(ns).Update(result)
+		return updateErr
+	})
+	if retryErr != nil {
+		glog.Errorf("Update of PVC %s/%s failed: %s", ns, pvc, retryErr)
+		return retryErr
+	}
+	return nil
+}
+
+// mount the given PV if not mounted yet, return the mountpoint or an error
+func (p *glusterSubvolProvisioner) mountPV(ns string, pv *v1.PersistentVolume) (string, error) {
+	// check if not mounted yet
+	mountpoint, mounted := p.mtab[pv.Name]
+	if mounted {
+		// mounted already
+		return mountpoint, nil
+	}
+
+	// create the missing mountpoint
+	mountpoint = workdir + "/" + pv.Name
+	sb, err := os.Stat(mountpoint)
+	if err != nil {
+		// mountpoint does not exist yet, create it
+		glog.Infof("mountpoint %s for PV %s does not exist yet? Error: %s", mountpoint, pv.Name, err)
+
+		err = os.MkdirAll(mountpoint, 0775)
+		if err != nil {
+			return "", fmt.Errorf("failed to create mountpoint %s for PV %s: %s", mountpoint, pv.Name, err)
+		}
+	} else if !sb.Mode().IsDir() {
+		// mountpoint should be a directory, but it is not?!
+		return "", fmt.Errorf("mountpoint %s for PV %s is not a directory?", mountpoint, pv.Name)
+	}
+
+	// get the name of the supervol to mount
+	supervol := pv.Spec.Glusterfs.Path
+
+	// get the mount options for the supervol
+	var mountOpts string
+	if len(pv.Spec.MountOptions) != 0 {
+		mountOpts = strings.Join(pv.Spec.MountOptions, ",")
+	}
+
+	// Gluster Storage servers
+	ep, err := p.client.CoreV1().Endpoints(ns).Get(pv.Spec.Glusterfs.EndpointsName, metav1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("endpoint %s for PV %s count not be found: %s", pv.Spec.Glusterfs.EndpointsName, pv.Name, err)
+	}
+
+	// add the additional servers as backup-volfile-servers to the mountOpts
+	if len(ep.Subsets[0].Addresses) > 1 {
+		var backupVolfileServers string
+		for _, addr := range ep.Subsets[0].Addresses[1:] {
+			if backupVolfileServers != "" {
+				backupVolfileServers = backupVolfileServers + ":"
+			}
+			backupVolfileServers = backupVolfileServers + addr.IP
+		}
+
+		if mountOpts != "" {
+			mountOpts = mountOpts + ","
+		}
+		mountOpts = mountOpts + "backup-volfile-servers=" + backupVolfileServers
+	}
+
+	mountCmd := make([]string, 0, 7)
+	mountCmd = append(mountCmd, "/bin/mount", "-t", "glusterfs")
+
+	// don't forget the "-o" option before the mount options string
+	if mountOpts != "" {
+		mountOpts = "-o " + mountOpts
+		mountCmd = append(mountCmd, "-o", mountOpts)
+	}
+
+	mountSource := fmt.Sprintf("%s:%s", ep.Subsets[0].Addresses[0].IP, supervol)
+
+	mountCmd = append(mountCmd, mountSource, mountpoint)
+	glog.Infof("going to mount supervol %s for PV %s: %s", supervol, pv.Name, mountCmd)
+
+	// os.StartProcess() does not allow deamons (fuse) to continue, using
+	// exec.Command instead.
+	cmd := exec.Command(mountCmd[0], mountCmd[1:]...)
```

Comment:
@nixpanic One logical question. Suppose supervol mount was successful or existed and just before subvol mount the supervol got unmounted. what happens for subvol PVC creation . It will be successful isnt it ? do you foresee any issues with it as the subvol mount is NOT on gluster mount ?

Reply:
I'm not sure what you are aiming at? There should not be a reason for the supervol to suddenly get unmounted...

Reply:
I am kind of suggesting, can we make sure we did a subvol creation always on supervol and not in local filesystem ?